src/runtime/components/forms/SelectMenu.vue

```diff
@@ -365,14 +365,35 @@ export default defineComponent({
 
     const selected = computed(() => {
       if (props.multiple) {
-        if (!Array.isArray(props.modelValue) || !props.modelValue.length) {
+        const modelValue = props.modelValue
+        if (!Array.isArray(modelValue) || !modelValue.length) {
           return []
         }
 
         if (props.valueAttribute) {
-          return options.value.filter(option => (props.modelValue as any[]).includes(option[props.valueAttribute]))
+          if (props.by) {
+            // handle the case when the valueAttribute is an object, and we need to compare by a specific key
+            return options.value.filter(
+              option => typeof option === 'object' && option !== null
+                && typeof option[props.valueAttribute] === 'object' && option[props.valueAttribute] !== null && modelValue.some(
+                  (value: any) => typeof value === 'object' && value !== null && value[props.by] === option[props.valueAttribute][props.by]
+                )
+            )
+          }
+
+          // compute selected items by the valueAttribute form the options
+          return options.value.filter(option => modelValue.includes(option[props.valueAttribute]))
```

Comment:
What if `modelValue` is an array of object? I think you should add a condition for `modelValue`. If `modelValue` is an object, then you should use `props.valueAttribute` in `modelValue`, for example: `modelValue[props.valueAttribute]`. Otherwise, you can compare it like in the current code. @CernyMatej

Reply:
I'm not really sure what you mean.
This code is executed when the `multiple` prop is set to `true`. In that case, you would always use an array for the `modelValue`, wouldn't you?
If there was an object first, it would change to an array after the user selects another item and I can't see any use case for that.

Reply:
I mean, the `modelValue` could be an array of objects, so I don't think you can use the `includes` method directly here, as it only works on arrays of `strings` or `numbers`.

Reply:
Sorry, I can't see the problem.
<img width="538" alt="image" src="https://github.com/user-attachments/assets/3869d21c-03fe-49f9-9113-2ee406885418">
<img width="622" alt="image" src="https://github.com/user-attachments/assets/250307d6-b73c-41e5-8c56-348ead3820e0">
![image](https://github.com/user-attachments/assets/88195253-b2f2-44ac-a73c-7916998be10f)

Reply:
> Sorry, I can't see the problem.
> 
> > then you should use props.valueAttribute in modelValue, for example: modelValue[props.valueAttribute]. Otherwise, you can compare it like in the current code.
> 
> How could I use `modelValue[props.valueAttribute]` if `modelValue` is an array?
> 
> <img alt="image" width="538" src="https://private-user-images.githubusercontent.com/112722215/383440805-3869d21c-03fe-49f9-9113-2ee406885418.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA4NzcxMjgsIm5iZiI6MTczMDg3NjgyOCwicGF0aCI6Ii8xMTI3MjIyMTUvMzgzNDQwODA1LTM4NjlkMjFjLTAzZmUtNDlmOS05MTEzLTJlZTQwNjg4NTQxOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwNlQwNzA3MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ODQzOWQ4ZDAwMTdmYmZmNmYyNWE3ZmQzMDdiNjZkMjhmYzg0YmRhMmVkOGU0MTQ2MGE5OGMwNDg0NzY3Zjg5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.59w4OTEaA9i_99JW6iyvDIkh9ONoHI4aXhbqhocmipc"> <img alt="image" width="622" src="https://private-user-images.githubusercontent.com/112722215/383441139-250307d6-b73c-41e5-8c56-348ead3820e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA4NzcxMjgsIm5iZiI6MTczMDg3NjgyOCwicGF0aCI6Ii8xMTI3MjIyMTUvMzgzNDQxMTM5LTI1MDMwN2Q2LWI3M2MtNDFlNS04YzU2LTM0OGVhZDM4MjBlMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwNlQwNzA3MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZjY0ZWMwNDI3MzE4YTY1ZDNmNTlmNTQzNDI1YzFmZGM2NTZjNjNhNmI2MDIyZmYzNjljMjgyNDM1YjlmMDBkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.i88vHrvZxw_JnJWwZC4riB3Fldn9BJvAuj0mIWo8YKc"> ![image](https://private-user-images.githubusercontent.com/112722215/383441692-88195253-b2f2-44ac-a73c-7916998be10f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA4NzcxMjgsIm5iZiI6MTczMDg3NjgyOCwicGF0aCI6Ii8xMTI3MjIyMTUvMzgzNDQxNjkyLTg4MTk1MjUzLWIyZjItNDRhYy1hNzNjLTc5MTY5OThiZTEwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwNlQwNzA3MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MjY5N2ExN2Y1OGMxYWU5NDc0ZjU2MTI4ZmMyOWQ5OTQ3ZGQ0YTU4ZTk4ZmRkZmEwOWQwNWE0MDc1OGQ5MGI2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.zCOmeoLv171Jaud58hgulAO0WUiyIyNmlRuP1M7W83I).

You can iterate over your `modelValue` using the array's `.some()` method. During iteration, you can check if each callback value is an object, and if so, use `valueAttribute` to access its `value` property.

Reply:
In picture number one:
It will be `true` because you're reusing the value `obj` inside options. This causes comparison by `reference`. You could try setting the initialValue of selected to:
```js
const selected = ref([
  {
    id: 1,
    title: "Hello",
  },
]);
```

Reply:
Sometimes in real-world cases, the initial value may not come from existing data, but instead may come from other resources.

Reply:
If the data accessed from modelValue[props.valueAttribute] is an object, we can temporarily use `JSON.stringify(obj1) === JSON.stringify(obj2)` for comparison. We can update this approach later when v3 is released since it will include the `fast-deep-equal` library.

Reply:
If the data accessed from modelValue[props.valueAttribute] is an object, we can temporarily use JSON.stringify(obj1) === JSON.stringify(obj2) for comparison. We can update this approach later when v3 is released since it will include the fast-deep-equal library.

Reply:
> In picture number one: It will be `true` because you're reusing the value `obj` inside options. This causes comparison by `reference`. You could try setting the initialValue of selected to:
> 
> ```js
> const selected = ref([
>   {
>     id: 1,
>     title: "Hello",
>   },
> ]);
> ```

I see. Yeah, that will return `false`. I did not consider this use case relevant since one should probably always use the `by` prop to compare objects.


> You can iterate over your modelValue using the array's .some() method. During iteration, you can check if each callback value is an object, and if so, use valueAttribute to access its value property.

I'm not sure I understand what you're suggesting. The `reference` of the object in the options will never be the same in `modelValue`, if it is a different object, will it?
In the example I provided, the data under `value-attribute` is an object.

Reply:
Do you think it is worth relying on unreliable comparison using `JSON.stringify`? In my opinion, we should recommend using the `by` prop for object comparison.

Reply:
> > In picture number one: It will be `true` because you're reusing the value `obj` inside options. This causes comparison by `reference`. You could try setting the initialValue of selected to:
> > ```js
> > const selected = ref([
> >   {
> >     id: 1,
> >     title: "Hello",
> >   },
> > ]);
> > ```
> 
> I see. Yeah, that will return `false`. I did not consider this use case relevant since one should probably always use the `by` prop to compare objects.
> 
> > You can iterate over your modelValue using the array's .some() method. During iteration, you can check if each callback value is an object, and if so, use valueAttribute to access its value property.
> 
> I'm not sure I understand what you're suggesting. The `reference` of the object in the options will never be the same in `modelValue`, if it is a different object, will it? In the example I provided, the data under `value-attribute` is an object.

If the data accessed from modelValue[props.valueAttribute] is an object, we can temporarily use JSON.stringify(obj1) === JSON.stringify(obj2) for comparison. We can update this approach later when v3 is released since it will include the fast-deep-equal library.

Reply:
> Do you think it is worth relying on unreliable comparison using `JSON.stringify`? In my opinion, we should recommend using the `by` prop for object comparison. There is always something unique in the options, otherwise it wouldn't make sense to have a select in the first place.

Yes, you're right. I think we need to update the documentation for comparing objects.

Reply:
Should I change anything then?

Reply:
After seeing this, I'm not sure the dot notation works, though.

Reply:
can u give me example ?

Reply:
The documentation says that you can use dot notation like `'user.name'` to access nested object properties. However, even when I check the component several revisions back, it doesn't seem that this was supported.
Am I missing something?

It is used to directly access the properties everywhere:

<img width="441" alt="image" src="https://github.com/user-attachments/assets/67797b8c-ed51-479f-8a33-c02aecfd314a">

Reply:
The documentation refers to the prop `option-attribute`

Reply:
Currently, `valueAttribute` doesn't support `dot.notation`. Can we improve it? this be done in the same PR

Reply:
I'll have a look at it later today 😊